multicore_runtime/middleware.py

```diff
@@ -0,0 +1,185 @@
+
+  This requires a single-threaded webserver, or for os.environ to be patched to
+  be thread-local.
+
+  Args:
+    app: The WSGI app to wrap.
+    frozen_environment: An iterable of (key, value) tuples that can be used to
+      populate os.environ with the initial state of the environment.
+      `tuple(os.environ.iteritems())` produces appropriate output.
+    frozen_user_env: An iterable of (key, value) tuples that can be used to
+      populate os.environ with environment variables specified in app.yaml.
+    frozen_env_config_env: An iterable of (key, value) tuples that can be
+      used to populate os.environ with configuration-dependent env variables.
+
+  Returns:
+    The wrapped app, also a WSGI app.
+  """
+
+  def reset_environment_wrapper(wsgi_env, start_response):  # pylint: disable=missing-docstring
+    # Wipe os.environ entirely.
+    os.environ.clear()
+
+    # Populate os.environ in order, so that later steps overwrite conflicting
+    # keys in previous steps.
+
+    # Add in user env variables specified in app.yaml. These are added first
+    # because they should not take precedence over any conflicting key.
+    os.environ.update(frozen_user_env)
+
+    # Repopulate os.environ with the frozen environment.
+    os.environ.update(frozen_environment)
+
+    # Add in wsgi_env data, including request headers.
+    os.environ.update(request_environment_for_wsgi_env(wsgi_env))
+
+    # Add in configuration data from env_config.
+    os.environ.update(frozen_env_config_env)
+
+    # Add reserved keys, which draw from wsgi_env as well. These have a very
+    # high priority and so are added nearly last.
+    os.environ.update(reserved_env_keys_for_wsgi_env(wsgi_env))
+
+    # Tweak the environment to hide the service bridge.
+    os.environ.update(get_env_to_hide_service_bridge(wsgi_env))
+
+    return app(wsgi_env, start_response)
+
+  return reset_environment_wrapper
+
+
+def request_environment_for_wsgi_env(wsgi_env):
+  """Get a dict of key-value pairs from wsgi_env that work as env variables.
+
+  Does not mutate input.
+
+  Args:
+    wsgi_env: WSGI request env data.
+
+  Returns:
+    A dictionary suitable for use in `os.environ.update(output)`.
+  """
+
+  # Return all key, value pairs from wsgi_env where value is a string.
+  return {key: value for key, value in wsgi_env.iteritems()
+          if isinstance(value, basestring)}
+
+
+def reserved_env_keys_for_wsgi_env(wsgi_env):
+  """Get a dict for reserved keys based on wsgi_env headers and defaults.
+
+  Does not mutate input.
+
+  Args:
+    wsgi_env: WSGI request env data.
+
+  Returns:
+    A dictionary suitable for use in `os.environ.update(output)`.
+  """
+
+  output = {}
+
+  # Use the default value for a reserved key if the corresponding header is not
+  # set, or if the header exists but its value is blank.
+  for key, default in RESERVED_ENV_KEYS.iteritems():
+    value = wsgi_env.get('HTTP_X_APPENGINE_' + key)
```

Comment:
Usual recommendation is to use the `%` operator or `.format()` instead of `+` for string concatenation.
https://google-styleguide.googlecode.com/svn/trunk/pyguide.html?showone=Strings#Strings

(In this case it's probably fine as is, the problem cases are when there's multiple `+`s [which is inefficient] or cases where the variable might hold something that it's not a string, but it's good to get into the habit of always using those operators for string formatting.)